File: experiments/runs/2026-09-20_07-33-00_claim-office-java-example-mapping_exact-ptdd-v1-cc_opus-5-no-thinking/src/main/java/Policy.java
import java.util.ArrayList;
import java.util.List;

/** An insurance policy covering a list of items. */
final class Policy {

    private static final int CAP_FACTOR = 2;

    private final List<Item> items;
    private final int premium;
    private int remainingCap;

    Policy(List<Item> items, int premium) {
        this.items = List.copyOf(items);
        this.premium = premium;
        this.remainingCap = CAP_FACTOR * insuranceSum();
    }

    int premium() {
        return premium;
    }

    /**
     * Matches each reported damage to a distinct insured item. A damage whose
     * item is not covered by the policy is rejected, as is a second damage to
     * a type the policy insures only once.
     */
    List<DamagedItem> damagedItems(List<Damage> damages) {
        List<Item> unclaimed = new ArrayList<>(items);
        List<DamagedItem> damaged = new ArrayList<>();
        for (Damage damage : damages) {
            damaged.add(new DamagedItem(claimOne(unclaimed, damage), requireReportable(damage)));
        }
        return damaged;
    }

    private int requireReportable(Damage damage) {
        if (damage.amount() < 0) {
            throw new IllegalArgumentException(
                    "a damage amount cannot be negative: " + damage.amount());
        }
        return damage.amount();
    }

    private Item claimOne(List<Item> unclaimed, Damage damage) {
        for (int i = 0; i < unclaimed.size(); i++) {
            if (unclaimed.get(i).type().equals(damage.itemType())) {
                return unclaimed.remove(i);
            }
        }
        throw new IllegalArgumentException(
                "damaged item is not covered by the policy: " + damage.itemType());
    }

    int remainingCap() {
        return remainingCap;
    }

    /** The total payout per policy is capped at twice the insurance sum. */
    int payOutUpToCap(int desiredPayout) {
        int payout = Math.min(desiredPayout, remainingCap);
        remainingCap -= payout;
        return payout;
    }

    /** The insurance sum is the sum of the insured items' insurance values. */
    private int insuranceSum() {
        int sum = 0;
        for (Item item : items) {
            sum += PriceList.insuranceValueOf(item);
        }
        return sum;
    }
}

File: experiments/runs/2026-09-23_16-10-32_claim-office-example-mapping_exact-ptdd-v1.1-refactor-subagent-cc_opus-5-5-no-thinking/src/policy.ts
import type { InsuredItem } from "./insured-item";
import { unitInsuranceValue } from "./price-list";

// A policy's payout cap: the total payout per policy is capped at twice the insurance sum.
const CAP_FACTOR = 2;

export interface Policy {
  items: InsuredItem[];
  remainingCap: number;
}

// The insurance sum uses each item's price-list insurance value; premium-only rules such as curse surcharges
// or the building-block discount do not change it.
function insuranceSum(items: InsuredItem[]): number {
  return items.reduce((sum, item) => sum + unitInsuranceValue(item.type), 0);
}

export function openPolicy(items: InsuredItem[]): Policy {
  return { items, remainingCap: CAP_FACTOR * insuranceSum(items) };
}

// Pays out at most what is left of the policy's cap and deducts the payout from it.
export function drawFromCap(policy: Policy, payout: number): number {
  const cappedPayout = Math.min(payout, policy.remainingCap);
  policy.remainingCap -= cappedPayout;
  return cappedPayout;
}
